### PR TITLE
kafka replay speed: compressed/uncompressed bytes for MaxBytes in fetch requests

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
Normally all kafka flavours support the same API, except when they don't. Warpstream treats the `MaxBytes` field of a Fetch request as a limit on the uncompressed bytes of all the records. Apache Kafka treats it as a limit on the compressed bytes.

This difference in behaviour is quite significant and leads to very suboptimal performance when using the wrong interpretation.

This PR adds a config flag to be able to choose between the Kafka flavours. Open to feedback on what to call the flag and what to call the options.

#### Note to reviewers
 
This PR is based on #9203, but will also conflict with #9153 when merged. The conflicts should be tiny, so I think this is ready for a review. I'll rebase when the two PRs above are merged.

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by seizethedave** [comment]
Change looks good, but: in this route we end up encoding facts about each Kafka variant into the Mimir code. Since I expect these differences will be rare, I wondered if you considered just exposing config knobs to handle each difference (like `ingest-storage.kafka.track-compressed-bytes`.) These behaviors may also change over time per Kafka flavor as new versions come out.

**Comment by pracucci** [comment]
> I pushed a change to use a specific flag instead of a `flavour`.
> 
> There are now also some basic docs. I think we have to add significantly more docs before this feature is ready for use, so right now I added only the minimum necessary related to the compressed bytes flag.

I would have preferred the flavor way, to make it trivially easy to configure (at the end, we currently want to support only Kafka and Warpstream), but I can see different opinions on it so I will disagree and commit to the specific flag option. It's something we can always revisit, until the feature is experimental.

File: docs/sources/mimir/configure/configuration-parameters/index.md
Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:3838]:
```suggestion
  # When enabled, the fetch request MaxBytes field is computed using the
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (tacole02) [inline docs/sources/mimir/configure/configure-kafka-backend.md:10]:
```suggestion
# Configure the Grafana Mimir Kafka backend
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (tacole02) [inline docs/sources/mimir/configure/configure-kafka-backend.md:18]:
```suggestion
To set up Mimir to work with different Kafka backends, you need to configure some parameters.
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (tacole02) [inline docs/sources/mimir/configure/configure-kafka-backend.md:19]:
```suggestion
Here are the Kafka flavors and additional configurations needed to set them up in Mimir.
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (tacole02) [inline docs/sources/mimir/configure/configure-kafka-backend.md:23]:
```suggestion
Use the default options with Apache Kafka. No additional configuration is needed.
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (tacole02) [inline docs/sources/mimir/configure/configure-kafka-backend.md:27]:
```suggestion
Use the default options with Confluent Kafka. No additional configuration is needed.
```

File: pkg/storage/ingest/config.go
Comment (tacole02) [inline pkg/storage/ingest/config.go:138]:
```suggestion
	f.BoolVar(&cfg.UseCompressedBytesAsFetchMaxBytes, prefix+".use-compressed-bytes-as-fetch-max-bytes", true, "When enabled, the fetch request MaxBytes field is computed using the compressed size of previous records. When disabled, MaxBytes is computed using uncompressed bytes. Different Kafka implementations interpret MaxBytes differently.")
```

File: pkg/storage/ingest/reader.go
Comment (tacole02) [inline pkg/storage/ingest/reader.go:830]:
```suggestion
	// trackCompressedBytes controls whether to calculate MaxBytes for fetch requests based on previous responses' compressed or uncompressed bytes.
```

**Review by tacole02 — APPROVED** [review]
Docs look good!

File: cmd/mimir/help-all.txt.tmpl
Comment (dimitarvdimitrov) [inline cmd/mimir/help-all.txt.tmpl:1359]:
```suggestion
    	When enabled, the fetch request MaxBytes field is computed using the compressed size of previous records. When disabled, MaxBytes is computed using uncompressed bytes. Different Kafka implementations interpret MaxBytes differently. (default true)
```

File: cmd/mimir/help.txt.tmpl
Comment (dimitarvdimitrov) [inline cmd/mimir/help.txt.tmpl:435]:
```suggestion
    	When enabled, the fetch request MaxBytes field is computed using the compressed size of previous records. When disabled, MaxBytes is computed using uncompressed bytes. Different Kafka implementations interpret MaxBytes differently. (default true)
```

File: docs/sources/mimir/configure/configure-kafka-backend.md
Comment (dimitarvdimitrov) [reply docs/sources/mimir/configure/configure-kafka-backend.md:10]: @tacole02 I think this doesn't recognize Kafka as a product name. Do you know where we need to make a change for the linter to expect "Kafka" and not "kafka"?

Comment (tacole02) [reply docs/sources/mimir/configure/configure-kafka-backend.md:10]: That's a good question, @dimitarvdimitrov , I'll defer to @jdbaldry.

Comment (dimitarvdimitrov) [reply docs/sources/mimir/configure/configure-kafka-backend.md:10]: i'll merge this now because it's still not going to the `main` branch and I need the change. But we should figure this out before we merge to `main` (or go with "kafka" instead of "Kafka")

**Pull request merged**
